### PR TITLE
Merkle tree pruning

### PR DESCRIPTION
**Pull request by SupremoUGH**
Implements Merkle tree pruning and nonprovable insertion.
Goes together with: https://github.com/Manta-Network/sdk/pull/118

Misc: 
- fixes a bug with the initial sync where the loop would be stuck because the checkpoint wasn't getting updated.
- improves the signer key logic

---

Before we can merge this PR, please make sure that all the following items have been checked off:

- [x] Linked to an issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Added **one** line describing your change in [`CHANGELOG.md`](https://github.com/manta-network/manta-rs/blob/main/CHANGELOG.md) and added the appropriate `changelog` label to the PR.
- [x] Re-reviewed `Files changed` in the GitHub PR explorer.
- [x] Checked that changes and commits conform to the standards outlined in [`CONTRIBUTING.md`](https://github.com/manta-network/manta-rs/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: manta-accounting/src/wallet/mod.rs [hunk 1]
```diff
@@ -339,7 +339,7 @@ where
             let ReadResponse {
                 should_continue,
                 data,
-            } = self.read_from_ledger().await?;
+            } = self.read_from_ledger(&checkpoint).await?;
```
Comment (zqhxuyuan) [inline manta-accounting/src/wallet/mod.rs:342]: the checkpoint here at first will be Default(0), so every time the wallet trying to call `initial_sync`, it will sync from beginning? So I suppose this initial_sync is only called once when user first time install the wallet. What if usre disable the wallet and enable it again, do we call initial_sync again? @DanielZhangReal

Comment (DanielZhangReal) [reply manta-accounting/src/wallet/mod.rs:342]: It will only be called once, and the result will be saved to indexedDB after success. 
Disabling and enabling will not affect indexedDB. 
Also, this logic only exists when the user creates a wallet, not the import logic

**Review by zqhxuyuan — DISMISSED** [review]

File: manta-pay/src/simulation/mod.rs [hunk 2]
```diff
@@ -72,7 +72,6 @@ where
         utxo_accumulator_model,
     );
     signer.load_accounts(AccountTable::new(KeySecret::sample(rng)));
-    signer.update_authorization_context();
```
Comment (bhgomes) [inline manta-pay/src/simulation/mod.rs:75]: Why was this removed?

File: manta-crypto/src/merkle_tree/partial.rs [hunk 3]
```diff
@@ -365,12 +377,64 @@ where
         self.push_leaf_digest(parameters, Node(len), leaf_digest()?);
         Some(true)
     }
+
+    /// Removes the paths corresponding to the nonprovable leaves in `self`.
+    #[inline]
+    pub fn prune(&mut self) {
+        if let Some(current_index) = self.leaf_map.current_index() {
+            let marked_indices = (0..current_index)
```
Comment (bhgomes) [inline manta-crypto/src/merkle_tree/partial.rs:385]: Maybe leave a comment on why you have to collect before looping.

**Review by bhgomes — CHANGES_REQUESTED** [review]

File: manta-pay/src/simulation/mod.rs (hunk 2 above)
Comment (SupremoUGH) [reply manta-pay/src/simulation/mod.rs:75]: `load_accounts` now calls `update_authorization_context` internally.

**Review by bhgomes — APPROVED** [review]

**Review by zqhxuyuan — APPROVED** [review]
